articles/build-a-restful-api-with-adonisjs/index.md

```diff
@@ -0,0 +1,419 @@
+
+Read through the instructions and update your `.env` file with your database credentials as follows.
+
+```bash
+    DB_CONNECTION=mysql
+    MYSQL_USER= //DB_USER
+    MYSQL_HOST=localhost
+    MYSQL_DB_NAME= //DB_NAME
+    MYSQL_PORT=3306
+    MYSQL_PASSWORD= //DB_PASSWORD
+```
+
+If you want to change the default configurations for the databases, you can always go to `config/database.ts` to configure some credentials.
+
+If you encounter this error `Client does not support authentication protocol requested by server;` while testing your Forum API, follow these steps to solve it.
+
+```bash
+    npm install mysql2
+```
+
+Then open `config/database.ts` and update the `client` field to `mysql2`.
+
+That’s all.
+
+### Setting up Authorization and Authentication
+In AdonisJS 5, authentication and authorization is very easy to set up, all you need to do is to install the Auth package and every other complicated authentication logic is built in for you already.
+
+Let’s get started:
+
+Install the Auth package with this command.
+
+```bash
+    npm i @adonisjs/auth@alpha
+```
+
+As always, invoke the Auth package with the `invoke` command to configure it.
+
+```bash
+    node ace invoke @adonisjs/auth
+```
+
+First, it will ask you to select the provider, In this case, I chose Lucid and API Token next since we are building an API.
+
+1. Next, type in the `User` Model for your authentication,
+2. Then press `Y` key to create a migration for it.
+3. Next, choose `Database` as your provider
+4. Lastly, press `Y` again to create a `api_tokens` migration.
+
+Now, you should have 2 migrations in your `database/migrations` folder, update the `xxxxx_users.ts` file to include a `name` and any other columns of your choice.
+
+Lastly, add the `auth` middleware to the `kernel.ts` file inside `start/kernel.ts`:
+
+```ts
+Server.middleware.registerNamed({
+  auth: "App/Middleware/Auth",
+});
+```
+
+#### Creating migrations
+The next step is to create the remaining migrations for the Post and Forum models we will be creating later.
+
+Let’s get started:
+
+Create new migration using this command:
+
+```bash
+    node ace make:migration posts
+```
+
+After running the command, open the new file in `database/migrations/xxxx_posts.ts` and paste the code below.
+
+```ts
+import BaseSchema from "@ioc:Adonis/Lucid/Schema";
+export default class Posts extends BaseSchema {
+  protected tableName = "posts";
+  public async up() {
+    this.schema.createTable(this.tableName, (table) => {
+      table.increments("id");
+      table.string("title", 255).notNullable();
+      table.string("content", 255).notNullable();
+      table.integer("user_id", 180).notNullable();
+      table.integer("forum_id").nullable();
+      table.timestamps(true);
+    });
+  }
+  public async down() {
+    this.schema.dropTable(this.tableName);
+  }
+}
```

Comment:
Could you explain more about the above code.

Reply:
The code is used to create a database schema with the necessary Columns.

Reply:
done